Don't call `resetKeepAlive` in `connectionDidReceiveData`.

mattgallagher opened the pull request:
Hi Pawel,

As I mentioned in my other PR, I'm investigating why our server is choosing to close socket connections periodically.

I'm trying an experiment where I've removed the call to `resetKeepAlive` in `connectionDidReceiveData`. At the moment, it appears that this change avoids the problem I was having.

I don't know if this is correct. I don't really know what communication the SignalR server is supposed to use to keep the connection alive. I'm not even totally sure how our server is configured (it's out of my control). But I wanted to let you know the result of my experiment.

Cheers,
Matt.

moozzyk commented:
The idea of these keep-alive pings is to enable server to quickly detect when the client goes away - an example would be #159.

The behavior you are seeing could be explained by keep-alive pings - if the server does not hear from the client within the ClientTimeoutInterval (see [Configure Server Options](https://docs.microsoft.com/en-us/aspnet/core/signalr/configuration?view=aspnetcore-6.0&tabs=dotnet#configure-server-options)) it will close the connection assuming that the client went away. (Note that keep-alive messages are only sent for the webSocket transport. If you happen to be using long polling this should not be an issue.) AFAIU, the server will not enable this behavior until it receives the first ping. This client had no support for keep-alive until "[recently](#190)" - initially this keep-alive was not supported and then I accepted a PR that implemented this feature but it caused problems so, I disabled it by default (e.g. the latest version available on CocoaPods has this disabled by default). This brings me to the possible options for you: 
- you can disable keep-alive by setting the [keepAliveInterval](https://github.com/moozzyk/SignalR-Client-Swift/blob/48942fbe1bc41410d2550875d672caf29d7859d2/Sources/SignalRClient/HubConnectionOptions.swift#L17) to nil - it will do what you are trying to do in this PR
- you can try decreasing the default [keepAliveInterval](https://github.com/moozzyk/SignalR-Client-Swift/blob/48942fbe1bc41410d2550875d672caf29d7859d2/Sources/SignalRClient/HubConnectionOptions.swift#L17) since you have no control over the server. 

From my perspective it would be interesting to see the client logs and understand the timing - I am curious if the problem is that the server closes the connection because maybe it expects the pings more frequently or there is an issue on the client where the client stops sending keep-alive messages as per configured interval which in turn prompts the server to close the connection.

mattgallagher commented:
I looked at the log files and I understand what you are saying... my change disabled keep alive, entirely...

[LogWithKeepAliveAccidentallyDisabled.txt](https://github.com/moozzyk/SignalR-Client-Swift/files/8933303/LogWithKeepAliveAccidentallyDisabled.txt)

However, if I revert, the problem remains that the server is dropping the connection – 3 times in 3 minutes. Some keep alive are sent (you can see them in the log) but not many.

[LogWithKeepAliveAndConnectionLoss.txt](https://github.com/moozzyk/SignalR-Client-Swift/files/8933304/LogWithKeepAliveAndConnectionLoss.txt)

I suspect that my initial idea was correct (don't reset the keep alive when receiving data) but my implementation was wrong (I didn't realise that calling `resetKeepAlive` after handshake was required).

I have rebased this PR and changed the implementation. This new implementation keeps the call to `resetKeepAlive` but moves it up to immediately after the call to `connectionDidOpen`, so it is called once after the handshake but is not called in `connectionDidReceiveData` for the remainder of the connection (it is still called in `send` and when the keep alive succeeds).

[LogWithKeepAliveMoveToAfterHandshake.txt](https://github.com/moozzyk/SignalR-Client-Swift/files/8933311/LogWithKeepAliveMoveToAfterHandshake.txt)

This log shows the expected behavior... keep alive is sent and there are no connection failures (I ran the test for a few hours and the connection was lost only when my computer went to sleep).

moozzyk commented:
I think I figured out what is happening. After looking at your logs I realized it is a bit worse that I initially thought. The reason to keep resetting keep alive is to reduce chattiness. The problem is that the code as implemented now confused the server side keep-alives with the client side keep-alives. Keep-alive should not be reset on a receive but on a successful send. This explains why you see the issues - if the client is not sending anything but the server is, the client will keep resetting keep-alives. Now, since the client is not sending any real data nor keep-alives (as they are being reset) the server considers client to be dead and closes the connection. After I looked at your logs I realized this is actually a bigger issue. The problem is that even if your server code doesn't send anything to the client, the server does send - it continuously sends Pings (MessageType 6 messages which are keep-alive messages) to the client. Currently they will also reset keep alive on the client. Here is an example: 

```
2022-06-18 16:38:07.562023+1000 SignalRTestApp[91828:9911002] [Socket] Received data from transport | SignalRTestAppServices/SignalRClientService.swift:116
2022-06-18 16:38:07.562183+1000 SignalRTestApp[91828:9910790] [Socket] Data received | SignalRTestAppServices/SignalRClientService.swift:116
2022-06-18 16:38:07.562269+1000 SignalRTestApp[91828:9910790] [Socket] Reset keep alive | SignalRTestAppServices/SignalRClientService.swift:116
2022-06-18 16:38:07.562411+1000 SignalRTestApp[91828:9910790] [Socket] Payload contains 1 message(s) | SignalRTestAppServices/SignalRClientService.swift:116
2022-06-18 16:38:07.562478+1000 SignalRTestApp[91828:9910790] [Socket] Message received: {"type":6} | SignalRTestAppServices/SignalRClientService.swift:116
2022-06-18 16:38:22.567567+1000 SignalRTestApp[91828:9911569] [Socket] Received data from transport | SignalRTestAppServices/SignalRClientService.swift:116
2022-06-18 16:38:22.567764+1000 SignalRTestApp[91828:9910790] [Socket] Data received | SignalRTestAppServices/SignalRClientService.swift:116
2022-06-18 16:38:22.567869+1000 SignalRTestApp[91828:9910790] [Socket] Reset keep alive | SignalRTestAppServices/SignalRClientService.swift:116
2022-06-18 16:38:22.567995+1000 SignalRTestApp[91828:9910790] [Socket] Payload contains 1 message(s) | SignalRTestAppServices/SignalRClientService.swift:116
2022-06-18 16:38:22.568092+1000 SignalRTestApp[91828:9910790] [Socket] Message received: {"type":6} | SignalRTestAppServices/SignalRClientService.swift:116
2022-06-18 16:38:37.567722+1000 SignalRTestApp[91828:9911590] [Socket] Received data from transport | SignalRTestAppServices/SignalRClientService.swift:116
2022-06-18 16:38:37.568236+1000 SignalRTestApp[91828:9910790] [Socket] Data received | SignalRTestAppServices/SignalRClientService.swift:116
2022-06-18 16:38:37.568390+1000 SignalRTestApp[91828:9910790] [Socket] Reset keep alive | SignalRTestAppServices/SignalRClientService.swift:116
```
So, your fix correctly addresses the issue of incorrect resets - we want to kick off sending keep alives upon a successful handshake but don't want to reset keep alive when data is received from the server. The change does not address the chattiness issue - to fix that we should reset keep-alive on a successful send. The reason it works is that the server considers the client active if it receives any data (real data or keep-alive pings).

moozzyk commented:
I will take this change as the keep alive chattiness can be addressed separately.

moozzyk commented:
#242 should prevent sending unnecessary keep-alive messages
